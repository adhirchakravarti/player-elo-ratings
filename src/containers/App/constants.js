export const MATCHES_PATH = "/matches";
export const TIMESTAMP_PATH = "/lastUpdated";

export const GET_UPDATE_TIMESTAMP = "app/App/GET_UPDATE_TIMESTAMP";
export const GET_MATCHES = "app/App/GET_MATCHES";
export const GET_MATCHES_SUCCESS = "app/App/GET_MATCHES_SUCCESS";
export const CALCULATE_PLAYER_RATINGS = "app/App/CALCULATE_PLAYER_RATINGS";
export const CALCULATE_PLAYER_RATINGS_SUCCESS =
  "app/App/CALCULATE_PLAYER_RATINGS_SUCCESS";
export const CREATE_RATING_TABLE_DATA = "app/App/CREATE_RATING_TABLE_DATA";

export const START_POLLING = "app/App/START_POLLING";
export const STOP_POLLING = "app/App/STOP_POLLING";
export const POLL_DELAY = 20000;

export const ADD_NEW_MATCH = "app/App/ADD_NEW_MATCH";
export const SHOW_NOTIFICATION = "app/App/SHOW_NOTIFICATION";
export const HIDE_NOTIFICATION = "app/App/HIDE_NOTIFICATION";
