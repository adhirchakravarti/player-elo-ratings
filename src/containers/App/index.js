import React, { useEffect } from "react";
import PropTypes from "prop-types";
import {
  Switch,
  Route,
  Redirect,
  BrowserRouter as Router,
} from "react-router-dom";
import { connect, useDispatch } from "react-redux";
import { compose } from "redux";
import {
  createTableData,
  stopPolling,
  startPolling,
  hideNotification,
} from "./actions";

import Layout from "../../components/Layout";
import NotificationBar from "../../components/NotificationBar";

const mapStateToProps = (state) => ({
  matches: state.matches,
  lastUpdate: state.lastUpdate,
  playerRatingData: state.playerRatingData,
  ratingTableColumns: state.ratingTable.columns,
  ratingTableRowData: state.ratingTable.rowData,
  notification: state.notification,
  notificationType: state.notificationType,
  notificationMessage: state.notificationMessage,
});

function App({
  playerRatingData,
  ratingTableColumns,
  ratingTableRowData,
  notification,
  notificationType,
  notificationMessage,
}) {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(createTableData());
  }, [playerRatingData]);

  useEffect(() => {
    dispatch(startPolling());
    return () => dispatch(stopPolling());
  }, []);

  const handleSnackbarClose = () => {
    dispatch(hideNotification());
  };

  const message = {
    type: notificationType,
    message: notificationMessage,
  };

  return (
    <Router>
      <Switch>
        <Route exact path="/">
          <Redirect to="/ratings" />
        </Route>
        <Route path="/ratings">
          <>
            <Layout
              ratingTableColumns={ratingTableColumns}
              ratingTableRowData={ratingTableRowData}
              playerRatingData={playerRatingData}
            />
            <NotificationBar
              notification={notification}
              message={message}
              handleClose={handleSnackbarClose}
            />
          </>
        </Route>
      </Switch>
    </Router>
  );
}

App.propTypes = {
  matches: PropTypes.arrayOf(
    PropTypes.shape({
      createdAt: PropTypes.string,
      standings: PropTypes.arrayOf(PropTypes.string),
    })
  ),
  playerRatingData: PropTypes.object,
  ratingTableColumns: PropTypes.array,
  ratingTableRowData: PropTypes.array,
  notification: PropTypes.bool,
  notificationType: PropTypes.string,
  notificationMessage: PropTypes.string,
};

const withConnect = connect(mapStateToProps, null);

export default compose(withConnect)(App);
