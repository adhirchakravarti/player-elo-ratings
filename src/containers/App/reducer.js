import { original, produce, setAutoFreeze } from "immer";
setAutoFreeze(false);

import {
  GET_MATCHES,
  GET_MATCHES_SUCCESS,
  CALCULATE_PLAYER_RATINGS_SUCCESS,
  CREATE_RATING_TABLE_DATA,
  SHOW_NOTIFICATION,
  HIDE_NOTIFICATION,
} from "./constants";

export const initialState = {
  matches: [],
  lastUpdate: "",
  playerRatingData: {},
  ratingTable: {
    columns: [],
    rowData: [],
  },
  notification: false,
  notificationType: "",
  notificationMessage: "",
};

const rootReducer = produce((draft, action) => {
  switch (action.type) {
    case GET_MATCHES_SUCCESS: {
      const {
        matchData: { list: matches, updatedAt },
      } = action.payload;
      draft.matches = [...matches];
      draft.lastUpdate = updatedAt;
      break;
    }
    case CALCULATE_PLAYER_RATINGS_SUCCESS: {
      const { playerRatings } = action.payload;
      draft.playerRatingData = Object.assign({}, playerRatings);
      break;
    }
    case CREATE_RATING_TABLE_DATA: {
      const columns = [
        { title: "Player Name", field: "name" },
        { title: "Start Rating", field: "startRating", type: "numeric" },
        { title: "End Rating", field: "endRating", type: "numeric" },
        {
          title: "Rating Change",
          field: "change",
          type: "numeric",
        },
        {
          title: "Number of Matches",
          field: "matches",
          type: "numeric",
        },
      ];
      const { playerRatingData } = original(draft);
      const rowData = Object.keys(playerRatingData)
        .map((key) => {
          return {
            name: key,
            startRating: playerRatingData[key].preRating,
            endRating: playerRatingData[key].postRating,
            change: playerRatingData[key].change,
            matches: playerRatingData[key].matches.length,
          };
        })
        .filter((player) => player.matches >= 3);
      draft.ratingTable.columns = columns;
      draft.ratingTable.rowData = rowData;
      break;
    }
    case SHOW_NOTIFICATION: {
      const { notification } = action.payload;
      draft.notification = true;
      draft.notificationMessage = notification.message;
      draft.notificationType = notification.type;
      break;
    }
    case HIDE_NOTIFICATION: {
      draft.notification = false;
      draft.notificationMessage = "";
      draft.notificationType = "";
      break;
    }
  }
}, initialState);

export default rootReducer;
