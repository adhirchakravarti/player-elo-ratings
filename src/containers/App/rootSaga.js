import { spawn } from "redux-saga/effects";
import appSaga from "./saga";
import watchPollSaga from "./pollSaga";

function* rootSaga() {
  yield spawn(appSaga);
  yield spawn(watchPollSaga);
}

export default rootSaga;
