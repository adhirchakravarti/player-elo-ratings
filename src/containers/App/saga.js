import { call, put, takeLatest } from "redux-saga/effects";
import {
  MATCHES_PATH,
  GET_MATCHES,
  CALCULATE_PLAYER_RATINGS,
  ADD_NEW_MATCH,
} from "./constants";
import axiosGetRequest from "../../utils/axiosGetRequest";
import axiosPostRequest from "../../utils/axiosPostRequest";
import {
  getMatchesSuccess,
  calculatePlayerRatings,
  calculatePlayerRatingsSuccess,
  showNotification,
} from "./actions";
import ELOMatch from "../../utils/EloMatch";

function* getMatches() {
  const path = MATCHES_PATH;
  try {
    const response = yield call(axiosGetRequest, path);
    if (response instanceof Error) {
      throw response;
    }
    yield put(getMatchesSuccess(response));
    yield put(calculatePlayerRatings(response));
  } catch (error) {
    console.error(error);
    yield put(
      showNotification({
        message: `${error.toString()}`,
        type: "error",
      })
    );
  }
}

function* calculatePlayerELORatings(action) {
  try {
    const {
      matchData: { list: matches },
    } = action.payload;
    const playerData = {};
    matches.forEach((match) => {
      const eloM = new ELOMatch();
      const numberOfPlayers = match.standings.length;
      for (let i = 0; i < numberOfPlayers; i++) {
        const player = match.standings[i];
        const existingELOscore = playerData[player] !== undefined;
        if (existingELOscore) {
          eloM.addPlayer(player, i, playerData[player].postRating);
        } else {
          eloM.addPlayer(player, i, 1000);
        }
      }
      eloM.calculateELOs();
      eloM.getPlayers().forEach((player) => {
        if (!playerData[player.name]) {
          playerData[player.name] = {};
          playerData[player.name].preRating = player.eloPre;
          playerData[player.name].postRating = player.eloPost;
          playerData[player.name].change = player.eloChange;
          playerData[player.name].matches = [];
          playerData[player.name].matches.push({
            createdAt: match.createdAt,
            place: player.place,
            standings: match.standings,
            points: player.eloChange,
            rating: player.eloPost,
          });
        } else {
          playerData[player.name].change += player.eloChange;
          playerData[player.name].postRating += player.eloChange;
          playerData[player.name].matches.push({
            createdAt: match.createdAt,
            place: player.place,
            standings: match.standings,
            points: player.eloChange,
            rating: player.eloPost,
          });
        }
      });
    });
    yield put(calculatePlayerRatingsSuccess(playerData));
    yield put(
      showNotification({
        message: "Retrieved updates!",
        type: "success",
      })
    );
  } catch (error) {
    console.error(error);
    yield put(
      showNotification({
        message: `${error.toString()}`,
        type: "error",
      })
    );
  }
}

function* submitNewMatch(action) {
  const { match } = action.payload;
  const path = MATCHES_PATH;
  try {
    const response = yield call(axiosPostRequest, path, match);
    if (response instanceof Error) {
      throw response;
    }
    yield put(
      showNotification({
        message: "Generated a New Match!",
        type: "success",
      })
    );
  } catch (error) {
    console.error(error);
    yield put(
      showNotification({
        message: `${error.toString()}`,
        type: "error",
      })
    );
  }
}

export default function* appSaga() {
  yield takeLatest(GET_MATCHES, getMatches);
  yield takeLatest(CALCULATE_PLAYER_RATINGS, calculatePlayerELORatings);
  yield takeLatest(ADD_NEW_MATCH, submitNewMatch);
}
