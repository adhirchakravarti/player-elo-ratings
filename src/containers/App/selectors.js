import { initialState } from "./reducer";

const selectAppState = (state) => state || initialState;

const selectLastUpdate = (state) => state.lastUpdate;

export { selectAppState, selectLastUpdate };
