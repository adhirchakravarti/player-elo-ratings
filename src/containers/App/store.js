import { createStore, applyMiddleware, compose } from "redux";
import createSagaMiddleware from "redux-saga";
import rootSaga from "./rootSaga";
import rootReducer from "./reducer";

export const initialState = {
  matches: [],
  lastUpdate: "",
  playerRatingData: {},
  ratingTable: {
    columns: [],
    rowData: [],
  },
  notification: false,
  notificationType: "",
  notificationMessage: "",
};

const sagaMiddleware = createSagaMiddleware();

const composeEnhancers =
  typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    : compose;

const enhancer = composeEnhancers(applyMiddleware(sagaMiddleware));

const store = createStore(rootReducer, initialState, enhancer);
sagaMiddleware.run(rootSaga);

export default store;
