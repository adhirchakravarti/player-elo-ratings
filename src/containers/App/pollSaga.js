import { call, put, take, delay, race, select } from "redux-saga/effects";
import moment from "moment";
import axiosGetRequest from "../../utils/axiosGetRequest";
import {
  START_POLLING,
  STOP_POLLING,
  POLL_DELAY,
  TIMESTAMP_PATH,
} from "./constants";
import { stopPolling, getMatches, showNotification } from "./actions";
import { selectLastUpdate } from "./selectors";

function* workerPollSaga(action) {
  while (true) {
    try {
      const path = TIMESTAMP_PATH;
      const response = yield call(axiosGetRequest, path);
      if (response instanceof Error) {
        throw response;
      }
      const { lastUpdated: retrievedTimestamp } = response;
      const selectedTimeStamp = yield select(selectLastUpdate);
      if (
        selectedTimeStamp !== "" &&
        moment(retrievedTimestamp).valueOf() >
          moment(selectedTimeStamp).valueOf()
      ) {
        yield put(
          showNotification({
            message: "Update check - Update Available, Retrieving...",
            type: "info",
          })
        );
        yield put(getMatches());
      } else if (selectedTimeStamp === "") {
        yield put(
          showNotification({
            message: "Retrieving Matches...",
            type: "info",
          })
        );
        yield put(getMatches());
      } else {
        yield put(
          showNotification({
            message: "Update check - Up to date",
            type: "info",
          })
        );
      }
      yield delay(POLL_DELAY);
    } catch (error) {
      console.warn(error);
      yield put(
        showNotification({
          message: `${error.toString()}`,
          type: "error",
        })
      );
      yield put(stopPolling(error));
    }
  }
}

export default function* watchPollSaga() {
  while (true) {
    yield take(START_POLLING);
    yield race([call(workerPollSaga), take(STOP_POLLING)]);
  }
}
