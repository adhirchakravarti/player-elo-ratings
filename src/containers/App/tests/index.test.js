import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { ThemeProvider } from "@material-ui/core/styles";
import { StylesProvider } from "@material-ui/styles/";
import CssBaseline from "@material-ui/core/CssBaseline";
import GlobalResponsiveFont from "../../../globalResponsiveFontCSS";
import theme from "../../../theme";
import store from "../store";

import App from "../index";

function getPreparedPage(store) {
  const generateClassName = (rule, styleSheet) =>
    `${styleSheet.options.classNamePrefix}-${rule.key}`;
  return (
    <Provider store={store}>
      <StylesProvider generateClassName={generateClassName}>
        <ThemeProvider theme={theme}>
          <CssBaseline />
          <GlobalResponsiveFont />
          <MemoryRouter>
            <App />
          </MemoryRouter>
        </ThemeProvider>
      </StylesProvider>
    </Provider>
  );
}

describe("<App />", () => {
  xit("renders correctly and matches snapshot", () => {
    /* skipped because Material-UI generates random class counters at each render
    and that doesnt' match the class counter in the snapshot, tried below based on the
    following https://github.com/mui-org/material-ui/issues/9492 but it still doesn't work.
    I suspect that I should mock child components (need to learn how) */
    const { asFragment } = render(getPreparedPage(store));
    const initialRender = asFragment();
    expect(initialRender).toMatchSnapshot();
  });
  it("does not log errors in the console", () => {
    const spy = jest.spyOn(global.console, "error");
    render(getPreparedPage(store));
    expect(spy).not.toHaveBeenCalled();
  });
});
