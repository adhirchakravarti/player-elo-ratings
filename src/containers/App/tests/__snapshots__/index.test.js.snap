// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 MuiGrid-direction-xs-column MuiGrid-justify-xs-center MuiGrid-grid-xl-12"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          >
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton MuiIconButton-colorInherit MuiIconButton-edgeStart"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <span
                  aria-hidden="true"
                  class="material-icons MuiIcon-root"
                >
                  menu
                </span>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <h6
              class="MuiTypography-root MuiTypography-h6"
            >
              Scopely Player Ratings
            </h6>
          </div>
        </header>
        <div />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
      >
        <div
          class="MuiGrid-root makeStyles-tableContainer MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-10 MuiGrid-grid-lg-8 MuiGrid-grid-xl-8"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation2 MuiPaper-rounded"
              style="position: relative;"
            >
              <div
                class="MuiToolbar-root MuiToolbar-regular MTableToolbar-root MuiToolbar-gutters"
              >
                <div
                  class="MTableToolbar-title"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-h6"
                    style="white-space: nowrap; overflow: hidden; text-overflow: ellipsis;"
                  >
                    Player Ratings
                  </h6>
                </div>
                <div
                  class="MTableToolbar-spacer"
                />
                <div
                  class="MuiFormControl-root MuiTextField-root MTableToolbar-searchField"
                >
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd"
                  >
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                        focusable="false"
                        title="Search"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                        />
                      </svg>
                    </div>
                    <input
                      aria-invalid="false"
                      aria-label="Search"
                      class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
                      placeholder="Search"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                    >
                      <button
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-disabled MuiButtonBase-disabled"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="MuiIconButton-label"
                        >
                          <span
                            aria-hidden="true"
                            class="material-icons MuiIcon-root MuiIcon-fontSizeSmall"
                          >
                            clear
                          </span>
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
                <div
                  class="MTableToolbar-actions"
                >
                  <div>
                    <div>
                      <span>
                        <button
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
                          tabindex="0"
                          title="Add a New Random Match"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              add
                            </span>
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="Component-horizontalScrollContainer"
                style="overflow-x: auto; position: relative;"
              >
                <div>
                  <div
                    style="overflow-y: auto;"
                  >
                    <div>
                      <table
                        class="MuiTable-root"
                        style="table-layout: auto;"
                      >
                        <thead
                          class="MuiTableHead-root"
                        >
                          <tr
                            class="MuiTableRow-root MuiTableRow-head"
                          >
                            <th
                              class="MuiTableCell-root MuiTableCell-head MTableHeader-header MuiTableCell-paddingCheckbox"
                              scope="col"
                              style="width: 48px; text-align: center; box-sizing: border-box;"
                            >
                              <span
                                aria-disabled="true"
                                class="MuiButtonBase-root MuiTableSortLabel-root MuiButtonBase-disabled"
                                role="button"
                                tabindex="-1"
                              >
                                Actions
                              </span>
                            </th>
                            <th
                              class="MuiTableCell-root MuiTableCell-head MTableHeader-header MuiTableCell-alignLeft"
                              scope="col"
                              style="box-sizing: border-box;"
                            >
                              <span
                                aria-disabled="false"
                                class="MuiButtonBase-root MuiTableSortLabel-root"
                                role="button"
                                tabindex="0"
                              >
                                <div
                                  aria-describedby="rbd-hidden-text-0-hidden-text-0"
                                  data-rbd-drag-handle-context-id="0"
                                  data-rbd-drag-handle-draggable-id="0"
                                  data-rbd-draggable-context-id="0"
                                  data-rbd-draggable-id="0"
                                  draggable="false"
                                  role="button"
                                  tabindex="0"
                                >
                                  Player Name
                                </div>
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                                  />
                                </svg>
                              </span>
                            </th>
                            <th
                              class="MuiTableCell-root MuiTableCell-head MTableHeader-header MuiTableCell-alignRight"
                              scope="col"
                              style="box-sizing: border-box;"
                            >
                              <span
                                aria-disabled="false"
                                class="MuiButtonBase-root MuiTableSortLabel-root"
                                role="button"
                                tabindex="0"
                              >
                                <div
                                  aria-describedby="rbd-hidden-text-0-hidden-text-0"
                                  data-rbd-drag-handle-context-id="0"
                                  data-rbd-drag-handle-draggable-id="1"
                                  data-rbd-draggable-context-id="0"
                                  data-rbd-draggable-id="1"
                                  draggable="false"
                                  role="button"
                                  tabindex="0"
                                >
                                  Start Rating
                                </div>
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                                  />
                                </svg>
                              </span>
                            </th>
                            <th
                              class="MuiTableCell-root MuiTableCell-head MTableHeader-header MuiTableCell-alignRight"
                              scope="col"
                              style="box-sizing: border-box;"
                            >
                              <span
                                aria-disabled="false"
                                class="MuiButtonBase-root MuiTableSortLabel-root"
                                role="button"
                                tabindex="0"
                              >
                                <div
                                  aria-describedby="rbd-hidden-text-0-hidden-text-0"
                                  data-rbd-drag-handle-context-id="0"
                                  data-rbd-drag-handle-draggable-id="2"
                                  data-rbd-draggable-context-id="0"
                                  data-rbd-draggable-id="2"
                                  draggable="false"
                                  role="button"
                                  tabindex="0"
                                >
                                  End Rating
                                </div>
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                                  />
                                </svg>
                              </span>
                            </th>
                            <th
                              class="MuiTableCell-root MuiTableCell-head MTableHeader-header MuiTableCell-alignRight"
                              scope="col"
                              style="box-sizing: border-box;"
                            >
                              <span
                                aria-disabled="false"
                                class="MuiButtonBase-root MuiTableSortLabel-root"
                                role="button"
                                tabindex="0"
                              >
                                <div
                                  aria-describedby="rbd-hidden-text-0-hidden-text-0"
                                  data-rbd-drag-handle-context-id="0"
                                  data-rbd-drag-handle-draggable-id="3"
                                  data-rbd-draggable-context-id="0"
                                  data-rbd-draggable-id="3"
                                  draggable="false"
                                  role="button"
                                  tabindex="0"
                                >
                                  Rating Change
                                </div>
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                                  />
                                </svg>
                              </span>
                            </th>
                            <th
                              class="MuiTableCell-root MuiTableCell-head MTableHeader-header MuiTableCell-alignRight"
                              scope="col"
                              style="box-sizing: border-box;"
                            >
                              <span
                                aria-disabled="false"
                                class="MuiButtonBase-root MuiTableSortLabel-root"
                                role="button"
                                tabindex="0"
                              >
                                <div
                                  aria-describedby="rbd-hidden-text-0-hidden-text-0"
                                  data-rbd-drag-handle-context-id="0"
                                  data-rbd-drag-handle-draggable-id="4"
                                  data-rbd-draggable-context-id="0"
                                  data-rbd-draggable-id="4"
                                  draggable="false"
                                  role="button"
                                  tabindex="0"
                                >
                                  Number of Matches
                                </div>
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                                  />
                                </svg>
                              </span>
                            </th>
                          </tr>
                        </thead>
                        <tbody
                          class="MuiTableBody-root"
                        >
                          <tr
                            class="MuiTableRow-root"
                            style="height: 490px;"
                          >
                            <td
                              class="MuiTableCell-root MuiTableCell-body"
                              colspan="6"
                              style="padding-top: 0px; padding-bottom: 0px; text-align: center;"
                            >
                              No records to display
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </div>
                  </div>
                </div>
              </div>
              <table
                class="MuiTable-root"
              >
                <tfoot
                  class="MuiTableFooter-root"
                  style="display: grid;"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-footer"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-footer MuiTablePagination-root Component-paginationRoot"
                      colspan="3"
                      icons="[object Object]"
                      style="float: right; overflow-x: auto;"
                    >
                      <div
                        class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar Component-paginationToolbar MuiToolbar-gutters"
                      >
                        <div
                          class="MuiTablePagination-spacer"
                        />
                        <p
                          class="MuiTypography-root MuiTablePagination-caption Component-paginationCaption MuiTypography-body2 MuiTypography-colorInherit"
                          id="mui-19533"
                        >
                          Rows per page:
                        </p>
                        <div
                          class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot Component-paginationSelectRoot"
                        >
                          <div
                            aria-haspopup="listbox"
                            aria-labelledby="mui-19533 mui-27154"
                            class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
                            id="mui-27154"
                            role="button"
                            tabindex="0"
                          >
                            <div
                              style="padding: 0px 5px;"
                            >
                              10 rows 
                            </div>
                          </div>
                          <input
                            aria-hidden="true"
                            class="MuiSelect-nativeInput"
                            tabindex="-1"
                            value="10"
                          />
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M7 10l5 5 5-5z"
                            />
                          </svg>
                        </div>
                        <p
                          class="MuiTypography-root MuiTablePagination-caption Component-paginationCaption MuiTypography-body2 MuiTypography-colorInherit"
                        >
                          0-0 of 0
                        </p>
                        <div
                          class="MTablePaginationInner-root"
                        >
                          <span
                            class=""
                            title="First Page"
                          >
                            <button
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-disabled MuiButtonBase-disabled"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </span>
                          <span
                            class=""
                            title="Previous Page"
                          >
                            <button
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-disabled MuiButtonBase-disabled"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </span>
                          <span
                            class="MuiTypography-root MuiTypography-caption"
                            style="flex: 1; text-align: center; align-self: center; flex-basis: inherit;"
                          >
                            0-0 of 0
                          </span>
                          <span
                            class=""
                            title="Next Page"
                          >
                            <button
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-disabled MuiButtonBase-disabled"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </span>
                          <span
                            class=""
                            title="Last Page"
                          >
                            <button
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-disabled MuiButtonBase-disabled"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <span
                                class="MuiIconButton-label"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="MuiSvgIcon-root"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <path
                                    d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                                  />
                                </svg>
                              </span>
                            </button>
                          </span>
                        </div>
                      </div>
                    </td>
                  </tr>
                </tfoot>
              </table>
              <div
                style="position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; z-index: 11;"
              >
                <div
                  style="display: table; width: 100%; height: 100%; background-color: rgba(255, 255, 255, 0.7);"
                >
                  <div
                    style="display: table-cell; width: 100%; height: 100%; vertical-align: middle; text-align: center;"
                  >
                    <div
                      class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                      role="progressbar"
                      style="width: 40px; height: 40px;"
                    >
                      <svg
                        class="MuiCircularProgress-svg"
                        viewBox="22 22 44 44"
                      >
                        <circle
                          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                          cx="44"
                          cy="44"
                          fill="none"
                          r="20.2"
                          stroke-width="3.6"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
